src/mame/drivers/tomsadvs.cpp

```diff
@@ -0,0 +1,110 @@
+|o  HD74LS144P  |_______________|   ______   |   |  |
+|o               ________________  |U6295|   |___|  |
+|o  _________   | EPROM         |  |_____|        o |
+|o  HD74LS144P  |_______________|                 o |
+|o               ________________                 o |
+|   _________   | EPROM         |   ______          |
+|.. HD74LS144P  |_______________|  |U6295|        o |
+|..   ________                     |_____|        o |
+|..  |ALTERA |           ____________             o |
+|..  |EPM7032LC4412     |U635H16BDC |        ____   |
+|..  |_______|          |___________|   ___  7805CT |
+| _________  _________  _____________  |  |       · |
+|HD74HC244P HD74LS139P | EPROM       | | <-HD74LS373P
+|                      |_____________| |__|       · |
+| _________  _________  _______________  ___  ___ · |
+||_DIPSx8_| |_DIPSx8_| |W87C32C       | Xtal HA17555|
+|                      |______________| 12.288MHz   |
+| _________  _________  _________  _________      o |
+||_DIPSx8_| |_DIPSx8_|  HD74LS273P HD74LS04P      o |
+|                       _________  _________      o |
+|                       HD74LS273P |ULN2003A|     o |
+| _________  _________  _________  _________  ____  |
+||ULN2003A| |ULN2003A| |ULN2003A|  HD74LS32P DS1307 |
+|                                              (__) |
+|  ......    .....    :::::::::::::::    ooo   BATT |
+|___________________________________________________|
+
+ And an aux I/O PCB with an EPM7032LC44-12.
+
+ Both machines share the same game mechanics as Taito's Ice Cold Beer (ICE probably got
+ some kind of license from Taito).
+
+*****************************************************************************************/
+
+#include "emu.h"
+#include "cpu/mcs51/mcs51.h"
+#include "sound/okim6295.h"
+#include "speaker.h"
+
+namespace
+{
+
+class tomsadvs_state : public driver_device
+{
+public:
+	tomsadvs_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag)
+		, m_maincpu(*this, "maincpu")
+	{
+	}
+
+	void tomsadvs(machine_config &config);
+
+protected:
+	required_device<mcs51_cpu_device> m_maincpu;
+};
+
+INPUT_PORTS_START(tomsadvs)
+INPUT_PORTS_END
+
+void tomsadvs_state::tomsadvs(machine_config &config)
+{
+	I80C32(config, m_maincpu, 12.288_MHz_XTAL); // Actually a Winbond W87C32C-40
+
+	SPEAKER(config, "lspeaker").front_left();
+	SPEAKER(config, "rspeaker").front_right();
+
+	okim6295_device &oki1(OKIM6295(config, "oki1", 12.288_MHz_XTAL/16, okim6295_device::PIN7_HIGH)); // Clock frequency & pin 7 not verified
+	oki1.add_route(ALL_OUTPUTS, "lspeaker", 0.45); // Guess
+	oki1.add_route(ALL_OUTPUTS, "rspeaker", 0.45); // Guess
+
+	okim6295_device &oki2(OKIM6295(config, "oki2", 12.288_MHz_XTAL/16, okim6295_device::PIN7_HIGH)); // Clock frequency & pin 7 not verified
+	oki2.add_route(ALL_OUTPUTS, "lspeaker", 0.45); // Guess
+	oki2.add_route(ALL_OUTPUTS, "rspeaker", 0.45); // Guessgit checkout 
```

Comment:
@clawgrip what did I say about checking your work **before** opening pull requests?  And @ajrhacker, haven’t I said enough times that people need to actually review pull requests before merging them?

Reply:
You're talking about the git command text on the comment that is obviously an unadverted error?

You think I haven't checked it two or three times? I could have checked it a hundred times and still fail to see it.

I drove 800 Km to dump those machines, had to stay at night leaving my wife and baby alone at home, but hey, there's a typo on a comment! If you want to discourage contributions, you're doing a perfect job.

I have some dyslexia and probably a problem with details, buy never been a problem on my whole career (more than 20 years) as a Java developer on big software companies (Sun Microsystems, Atos, Accenture, etc.)... Never but here, where I really feel afraid to contribute.
On every other projects every colleage is allways supportive, friendly, willing to help. Here I feel like my work is not appreciated at all, and that it's better to stop.

I'm like I am, if you feel like there's no place for me on this project, just tell me.

On the last PR there was an obvious typo on the PR description... You edited the PR to remove the name of the person who was all night checking against dozens of PCBs looking for errors to fix on MAME, but did not edited it for fixing the typo.
What was the point? To make sure this guy will never contribute again by removing his name (I'm talking about the PR description, not the whatsnew!!!) and tell everyone I make typos??

Reply:
>You're talking about the git command text on the comment that is obviously an unadverted error?
>
>You think I haven't checked it two or three times? I could have checked it a hundred times and still fail to see it.

OK, so if it would be missed after checking hundreds of times, why did I see it straight away?  It stands out visually, it’s highlighted in a spell-cheking editor, and if you simply read the comments in the file, surely you’d notice it.

>I drove 800 Km to dump those machines, had to stay at night leaving my wife and baby alone at home, but hey, there's a typo on a comment! If you want to discourage contributions, you're doing a perfect job.

Family comes first.  If this is taking you away from your family and causing a problem, then adjust your priorities,  Dumping PCBs isn’t more important than family.

>I have some dyslexia and probably a problem with details, buy never been a problem on my whole career (more than 20 years) as a Java developer on big software companies (Sun Microsystems, Atos, Accenture, etc.)... Never but here, where I really feel afraid to contribute.
>On every other projects every colleage is allways supportive, friendly, willing to help. Here I feel like my work is not appreciated at all, and that it's better to stop.

Look at this from my point of view.  I put a massive amount of time and effort into this project.  Consider how much time I spent reviewing over 1,600 files for #9788 – I haven’t just skimmed the changes, I’ve looked for logic errors and potential pitfalls in every case.  Look at the feedback cycle on pull request like #9841 and #9773, or #9762 – note that I don’t just say “this won’t work” and close it, I have to explain *why* it won’t work each time.  While this is going on, I’m still trying to implement usability improvements (e.g. #9783), fix bugs (e.g. #9812), give other developers a hand when they ask (e.g. #9748), slowly chip away at our big pile of technical debt (e.g. #9884), and occasionally actually work on emulation (e.g. #9699).  On top of that, I try to help with issue triage over at MAME Testers, and helping out users with questions on reddit.

I’m feeling very burned out at the moment, it’s 7:45AM and I haven’t slept all night, but I’m still trying to catch up on what’s going on in MAME because if I don’t, I’m only making a rod for my own back later.

What *does* frustrate me is seeing the same pattern of errors from the same people.  I get frustrated at seeing pull requests where the code won’t build or validate, or where it obviously won’t work.  I get frustrated at pull requests being merged with apparently no review at all.  I get frustrated when people call absolutely any error a “typo” when they’re clearly not typographical errors.  I’m not asking for contributors to be perfect.  I’m just asking people to slow down, put a fair effort in and not make the same mistakes repeatedly.  Quality is everyone’s responsibility.

>On the last PR there was an obvious typo on the PR description... You edited the PR to remove the name of the person who was all night checking against dozens of PCBs looking for errors to fix on MAME, but did not edited it for fixing the typo.

Well, first of all, I don’t know what PR you’re even talking about.  The most recent PR I edited the title of was #9887 to make it more descriptive.  I didn’t remove any names.  I often edit pull request titles to make them more descriptive or to better match the usual format.  I don’t frequently edit pull request descriptions at all.

When you’re writing a pull request title, consider that some developer is going to be seeing it in a list of search results months or possibly years later when trying to track something down, most likely while tired and frustrated.  Pull request titles should be terse and descriptive so someone has a good chance of seeing what it does without having to read the description, or resort to looking at the diff itself.  If a pull request title doesn’t follow the format developers expect on a project, it takes longer to interpret, also increasing the workload.  It’s not a place for stuffing names for street cred, and while using joke in PR titles might seem like a good idea at the time, it’s unhelpful.  It’s completely redundant to include your own name in a pull request title, as the author’s name is already part of the pull request metadata.

We have a significant problem with pull request titles that aren’t descriptive, and cases of only describing the pull request in a comment, so even expanding the description isn’t enough to work out what it’s supposed to do (remember web browsers aren‘t the only way people view data – GitHub has APIs that work with other applications, too).  This is an example of a pull request title and description that isn’t particularly useful – it just says “New NOT_WORKING machines”.  When you see that in a list, it doesn’t give any clues about what the machines are, and even the PR description doesn’t mention that it adds a skeleton driver.

>What was the point? To make sure this guy will never contribute again by removing his name (I'm talking about the PR description, not the whatsnew!!!) and tell everyone I make typos??

I still don’t know what you’re actually referring to.  If you’re referring to the summarised pull request titles that appear in the release announcement, there’s a strict 132-character line length limit.  I need to aggressively edit the titles to make them terse enough to fit, as well as coming up with descriptive titles for pull requests that were submitted with unhelpful titles.  The pull request author is always included (that would be you if it was a pull request you opened), but other names may be omitted just to save space.

If I missed a typo in the summarised pull request titles, I’m sorry.  Every month I need to do summarise a large number of bug reports and pull requests, as well as sorting the additions and promotions, and it’s always under time pressure given how much goes into each release.  I do check it multiple times, and I try to get it into public version control at least 24 hours before a release in case other people want to review it (@0kmg pointed out a couple of editing errors for 0.244).